CLOUDSC2 TL and AD GPU port via Loki with SCC and SCC-HOIST transformations

piotrows opened the pull request:
This PR adds GPU-enabled SCC and SCC-HOIST  ports for CLOUDSC2 TL and AD.  Although all the options are working correctly with manual edit of Loki generated files, its final shape will be known after missing Loki functionalities are mitigated.`     Start  8: dwarf-cloudsc2-tl--loki-scc-serial

8: Test command: /home/napz/dwarf-p-cloudsc2-loki-ad-gpu/build/bin/dwarf-cloudsc2-tl-loki-scc
8: Environment variables:
8:  NV_ACC_CUDA_HEAPSIZE=9G
8:  OMP_NUM_THREADS=1
8: Test timeout computed to be: 1500
8:      NUMOMP    NGPTOT  #GP-cols     #BLKS    NPROMA tid# : Time(msec)  MFlops/s
8:           1     16384     16384         1        32    0 :       1614       405 @ rank#0:core#128
8:           1     16384     16384       512        32   -1 :       3371       194 : TOTAL @ rank#0
8:       1 x 1     16384     16384       512        32   -1 :       3371       194 : TOTAL
8:   TL Taylor test
8:                  Lambda   Result
8:                         1    2.166696691743514
8:                         2    2.080780660321741
8:                         3    1.000161708616709
8:                         4    1.000016146275012
8:                         5    1.000001614797514
8:                         6    1.000000161000666
8:                         7    1.000000249799620
8:                         8    1.000004866493911
8:                         9    1.000035670344458
8:                        10    1.000456200005707
8:     ==============================================
8:         TEST PASSED, penalty             0
8:     ==============================================
 8/12 Test  #8: dwarf-cloudsc2-tl--loki-scc-serial .........   Passed    6.59 sec
test 9
      Start  9: dwarf-cloudsc2-tl-loki-scc-hoist-serial

9: Test command: /home/napz/dwarf-p-cloudsc2-loki-ad-gpu/build/bin/dwarf-cloudsc2-tl-loki-scc-hoist
9: Environment variables:
9:  NV_ACC_CUDA_HEAPSIZE=19G
9:  OMP_NUM_THREADS=1
9: Test timeout computed to be: 1500
9:      NUMOMP    NGPTOT  #GP-cols     #BLKS    NPROMA tid# : Time(msec)  MFlops/s
9:           1     16384     16384         1        32    0 :       2126       307 @ rank#0:core#128
9:           1     16384     16384       512        32   -1 :       3006       217 : TOTAL @ rank#0
9:       1 x 1     16384     16384       512        32   -1 :       3006       217 : TOTAL
9:   TL Taylor test
9:                  Lambda   Result
9:                         1    2.166696691743514
9:                         2    2.080780660321741
9:                         3    1.000161708616709
9:                         4    1.000016146275012
9:                         5    1.000001614797514
9:                         6    1.000000161000666
9:                         7    1.000000249799620
9:                         8    1.000004866493911
9:                         9    1.000035670344458
9:                        10    1.000456200005707
9:     ==============================================
9:         TEST PASSED, penalty             0
9:     ==============================================`
1: Test command: /home/napz/dwarf-p-cloudsc2-loki-ad-gpu/build/bin/dwarf-cloudsc2-ad-loki-scc
11: Environment variables:
11:  NV_ACC_CUDA_HEAPSIZE=9G
11:  OMP_NUM_THREADS=1
11: Test timeout computed to be: 1500
11:      NUMOMP    NGPTOT  #GP-cols     #BLKS    NPROMA tid# : Time(msec)  MFlops/s
11:           1     16384     16384         1        32    0 :       3683       177 @ rank#0:core#128
11:           1     16384     16384       512        32   -1 :       4969       131 : TOTAL @ rank#0
11:       1 x 1     16384     16384       512        32   -1 :       4969       131 : TOTAL
11:   AD TEST
11:   The maximum error is     30.15455588244263
11:   times the zero of the machine.
11:     =============================
11:     =           TEST OK         =
11:     =============================
11/12 Test #11: dwarf-cloudsc2-ad--loki-scc-serial .........   Passed    6.11 sec
test 12
      Start 12: dwarf-cloudsc2-ad--loki-scc-hoist-serial

12: Test command: /home/napz/dwarf-p-cloudsc2-loki-ad-gpu/build/bin/dwarf-cloudsc2-ad-loki-scc-hoist
12: Environment variables:
12:  NV_ACC_CUDA_HEAPSIZE=9G
12:  OMP_NUM_THREADS=1
12: Test timeout computed to be: 1500
12:      NUMOMP    NGPTOT  #GP-cols     #BLKS    NPROMA tid# : Time(msec)  MFlops/s
12:           1     16384         0         0        32    0 :       1456         0 @ rank#0:core#128
12:           1     16384         0       512        32   -1 :       1861       351 : TOTAL @ rank#0
12:       1 x 1     16384         0       512        32   -1 :       1861       351 : TOTAL
12:   AD TEST
12:   The maximum error is     30.15455588244263
12:   times the zero of the machine.
12:     =============================
12:     =           TEST OK         =
12:     =============================
12/12 Test #12: dwarf-cloudsc2-ad--loki-scc-hoist-serial ...   Passed    3.80 sec

piotrows committed:
Corrections for GPU execution

piotrows committed:
2tl cleanup from mistakenly altered variable names

piotrows committed:
Merge branch 'napz-loki-ad-gpu' of github.com:ecmwf-ifs/dwarf-p-cloudsc2-tl-ad into napz-loki-ad-gpu

piotrows committed:
AD scc-hoist fixed

piotrows committed:
TL scc works on CPU

piotrows committed:
Merge branch 'napz-loki-ad-gpu' of github.com:ecmwf-ifs/dwarf-p-cloudsc2-tl-ad into napz-loki-ad-gpu

piotrows committed:
2ad cleanup, timings excluded for GPU runs

piotrows committed:
TL scc works, all intermediate variables in driver with IBL added and error norm computations cast at the end of the driver routine.

piotrows committed:
Extension of AD loki temporaries with IBL

piotrows committed:
AD and TL scc-hoist work on GPU

piotrows committed:
TL and AD work properly on GPU

mlange05 committed:
Merge branch 'develop' into napz-loki-ad-gpu

mlange05 commented on src/cloudsc2_tl_loki/error.F90:
```diff
@@ -2,16 +2,22 @@ MODULE ERROR_MOD
-    END IF
+    INTEGER(KIND=JPIM) :: JK, JL
+    DO JL=KIDIA, KFDIA
+      IF(JL == KIDIA) THEN
```
Note to self: Why is this used here? This completely goes against the idea of applying this norm over the block, no?

mlange05 committed:
TL-Loki: Use host-side utilty routine for norm computation and check

By moving this to a single routine, we can easily disable thi in the
loki config. The new routine computes the global norms for all fields
on device before assessing the V-shape of the norms progression.

mlange05 committed:
Loki: Re-instate GPU-compatible timers via Loki-CPP preprocessing

mlange05 committed:
Loki: Reverting declaration workarounds

Previos version of the Loki-SCC-hoist transformation would struggle
with splitting variables into separate declarations when dimesnions
are changed. This now works with the new recursive hoisting mechanic,
so we can revert these changes now.

mlange05 committed:
Loki-AD: Allocate intermediate norm arrays per NPROMA block

mlange05 reviewed (commented):
Ok, I think this is now ready. I've played through a bunch of changes that makes this a little neater, and there's one Loki PR that needs merging before we can merge this one here. This is now functional and gives the right result for certain TL configurations and large AD runs for benchmarking. 

For future work, there is likely some more GPU optimisation needed, but for now this should prove that it is indeed possible to satisfy the tests on device with Loki-SCC transformation recipes.

mlange05 reviewed (approved):
Ok, following the merge of the new SCC-hoist variant in upstream Loki, this is looking good to go now! For anyone trying this, we now require bleeding-edge latest Loki. And as mentioned before, further improvements and tweaks to the GPU variants are certainly possible and planned, but will done in a separate follow-up PR.

Going in now...
